aptos-move/mvhashmap/src/types.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_aggregator::delta_change_set::DeltaOp;
+use aptos_crypto::hash::HashValue;
+use std::sync::Arc;
+
+pub type TxnIndex = u32;
+pub type Incarnation = u32;
+pub type Version = (TxnIndex, Incarnation);
+
+/// Returned as Err(..) when failed to read from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVDataError {
+    /// No prior entry is found.
+    NotFound,
+    /// Read resulted in an unresolved delta value.
+    Unresolved(DeltaOp),
+    /// A dependency on other transaction has been found during the read.
+    Dependency(TxnIndex),
+    /// Delta application failed, txn execution should fail.
+    DeltaApplicationFailure,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVCodeError {
+    /// No prior entry is found.
+    NotFound,
+    /// A dependency on other transaction has been found during the read.
+    Dependency(TxnIndex),
+}
+
+/// Returned as Ok(..) when read successfully from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVDataOutput<V> {
+    /// Result of resolved delta op, always u128. Unlike with `Version`, we return
+    /// actual data because u128 is cheap to copy and validation can be done correctly
+    /// on values as well (ABA is not a problem).
+    Resolved(u128),
+    /// Information from the last versioned-write. Note that the version is returned
+    /// and not the data to avoid copying big values around.
+    Versioned(Version, Arc<V>),
+}
+
+/// Returned as Ok(..) when read successfully from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVCodeOutput<M, X> {
+    /// Arc to the executable corresponding to the latest module, and an indicator
+    /// whether the executable corresponds to a storage-version module.
+    Executable((Arc<X>, bool)),
```

Comment:
I feel like named struct is better suited here. Sth like
```
Executable {
    binary: Arc<X>,
    is_storage: bool,
}
```

The `is_storage` is probably a bad name but can probably provide ppl with better context.

Reply:
Fair point, will do

Reply:
bool had to be ExecutableDescriptor, which I think makes things cleaner. 
This output has to be adjusted for VM trait anyway, so I don't want to introduce too many intermediate types there. Also, we will likely adjust them as we fill in the implementations in the next PRs.